### ISSUE TITLE
Does graphql-errors support graphql-ruby 1.8.0pre / class-based API?

### ISSUE THREAD

**Issue by benkimball** [exAspArk/graphql-errors#9]
We're using graphql-ruby 1.8.0.pre9 and we've specified our types using the new [class-based API](https://github.com/rmosolgo/graphql-ruby/blob/1.8-dev/guides/schema/class_based_api.md) that is under development in that gem. I tried a simple installation of graphql-errors to handle validation errors:

    GraphQL::Errors.configure(OurSchema) do
      rescue_from Mongoid::Errors::Validations do |e|
        GraphQL::ExecutionError.new("haha")
      end
    end

Then in our `MutationType` I have:

    module Types
      class MutationType < BaseType
        field :createAward, Types::AwardType, null: false do
          description 'Create a new award'
          argument :awardInput, Types::Input::AwardInputType, required: true
        end

        def create_award(award_input:)
          Award.create! attrs_from_input(award_input)
        end

        ...
      end
    end

The `create_award` function is raising a `Mongoid::Errors::Validations` error, but the `rescue_from` block is not being invoked.

**Comment by exAspArk** [comment]
Hey @benkimball! Thanks for opening this issue!

Yeah, I didn't plan to support `graphql` gem version `1.8` until the stable version is released :) But I'll try to take a look.

**Comment by benkimball** [comment]
Awesome, thanks for the quick response! I'll also try digging around a bit to see if I can discover what changes might be needed.

**Comment by y310** [comment]
It seems that error handling for 1.8 has been changed https://github.com/rmosolgo/graphql-ruby/pull/1313

I found that using `context.add_error` works as workaround (not sure if it's appropriate way or not though)

```ruby
GraphQL::Errors.configure(Schema) do
  rescue_from ActiveRecord::RecordNotFound do |exception, obj, args, ctx|
    ctx.add_error(GraphQL::ExecutionError.new(exception.message))
  end
end
```

**Comment by joshteng** [comment]
Spent hours figuring this out! Documentation is lacking / not updated for V1.8. 
For those working through GraphQL tutorial from https://www.howtographql.com/graphql-ruby/6-error-handling/ and facing similar issues: (credit to @y310 )
```ruby
class Resolvers::CreateLink < GraphQL::Function
  argument :description, !types.String
  argument :url, !types.String

  type Types::LinkType

  def call(_obj, args, ctx)
    link = Link.new(
      description: args[:description],
      url: args[:url],
      user: ctx[:current_user]
    )
    link.save!
    link
  rescue ActiveRecord::RecordInvalid => err
    # on error, return an error:
    ctx.add_error(GraphQL::ExecutionError.new(link.errors.full_messages))
    # https://github.com/exAspArk/graphql-errors/issues/9
  end
end
```

**Comment by haizop** [comment]
@exAspArk  - Now that graphql 1.8 has been released, are you planning on updating this gem? Thanks.

**Comment by exAspArk** [comment]
I wrote some tests for GraphQL 1.7 and 1.8 to run them in parallel through CI https://github.com/exAspArk/graphql-errors/pull/11. There were 0 changes in the implementation but the tests pass, `graphql` gem version `1.8.4`.

Here is an example of a schema with mutation https://github.com/exAspArk/graphql-errors/blob/0b3184132b27969813fbbfd1a0ed3f592b966d35/spec/fixtures/schema.rb#L70-L150

Can anyone help me to write a test to reproduce the issue?

**Comment by y310** [comment]
I tested latest graphql-errors master with graphql-ruby 1.8.0.pre11 to 1.8.4.
Test result is
- 1.8.0.pre11 to 1.8.3: failed
- 1.8.4: passed

```
  1) GraphQL::Errors GraphQL integration rescues the first Post::NotFound error in the list
     Failure/Error: old_resolve_proc.call(object, arguments, context)

     RuntimeError:
               Failed to implement Post.id, tried:

               - `PostType#id`, which did not exist
               - `GraphQL::ExecutionError#id`, which did not exist
               - Looking up hash key `:id` or `"id"` on `Post is invalid`, but it wasn't a Hash

               To implement this field, define one of the methods above (and check for typos)
     # ./lib/graphql/errors.rb:28:in `call'
     # ./lib/graphql/errors.rb:28:in `block in instrument'
     # ./spec/graphql/errors_spec.rb:40:in `block (3 levels) in <top (required)>'
```

So, apparently, some changes for 1.8.4 fixed the error.

Also, I noticed that even with 1.8.3 or older version the test passes if I change the `rescue_from` handler with below.

```diff
diff --git a/spec/fixtures/schema.rb b/spec/fixtures/schema.rb
index 0fc211a..c4863eb 100644
--- a/spec/fixtures/schema.rb
+++ b/spec/fixtures/schema.rb
@@ -131,7 +131,7 @@ when '1_8'

   GraphQL::Errors.configure(Schema) do
     rescue_from Post::NotFound, Post::Invalid do |exception|
-      GraphQL::ExecutionError.new('Post is invalid')
+      raise GraphQL::ExecutionError.new('Post is invalid')
     end
```

Hope this helps.

**Comment by exAspArk** [comment]
Hey @y310, thanks a lot for looking into it! 

Seems like the old behavior was returned somewhere between the [v1.8.3 and v.1.8.4](https://github.com/rmosolgo/graphql-ruby/compare/v1.8.3...v1.8.4) changes (and there are a lot of them). I suspect it's related to this [line](https://github.com/rmosolgo/graphql-ruby/blob/3f71c9d67ba3de41cbdebe0b7066592fc59caca9/lib/graphql/schema/member/instrumentation.rb#L75) with new `result.is_a?(GraphQL::ExecutionError)` check:

```ruby
if ctx.skip == result || ctx.schema.lazy?(result) || result.nil? || result.is_a?(GraphQL::ExecutionError) || ctx.wrapped_object
  result
  ...
```

**Comment by exAspArk** [comment]
I'm going to close the issue since `graphql` Ruby gem returned the previous behavior (`return GraphQL::ExecutionError` instead of new `raise GraphQL::ExecutionError`).

TL;DR use the latest `graphql` `1.7.x` or `1.8.x` version.
